sway-lib-std/src/storage.sw

```diff
@@ -106,3 +109,189 @@ impl<K, V> StorageMap<K, V> {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger or equal to len, there is no item to remove
+        assert(index < len);
+
+        // gets the element before removing it, so it can be returned
+        let removed_element = get::<V>(sha256((index, __get_storage_key())));
+
+        // for every element in the vec with an index greater than the input index,
+        // shifts the index for that element down one
+        let mut count = index + 1;
+        while count < len {
+            // gets the storage location for the previous index
+            let key = sha256((count - 1, __get_storage_key()));
+            // moves the element of the current index into the previous index
+            store::<V>(key, get::<V>(sha256((count, __get_storage_key()))));
+            
+            count += 1;
+        }
+
+        // decrements len by 1
+        store(__get_storage_key(), len - 1);
+
+        removed_element
+    }
+
+    /// Removes the element at the specified index and fills it with the last element
+    /// Does not preserve ordering. Also returns the element
+    ///
+    /// # Arguments
+    /// 
+    /// * `index` - The index of the vec to remove the item from
+    ///
+    /// # Reverts
+    /// 
+    /// Reverts if index is larger or equal to length of the vec
+    #[storage(read, write)]
+    pub fn swap_remove(self, index: u64) -> V {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger or equal to len, there is no item to remove
+        assert(index < len);
+
+        let hash_of_to_be_removed = sha256((index, __get_storage_key()));
+        // gets the element before removing it, so it can be returned
+        let element_to_be_removed = get::<V>(hash_of_to_be_removed);
+
+        let last_element = get::<V>(sha256((len - 1, __get_storage_key())));
+        store::<V>(hash_of_to_be_removed, last_element);
+
+        // decrements len by 1
+        store(__get_storage_key(), len - 1);
+
+        element_to_be_removed
+    }
+
+    /// Inserts the value at the given index, moving the current index's value aswell as the following's
+    /// Up one index
+    ///
+    /// # WARNING 
+    ///
+    /// Expensive for larger vecs
+    ///
+    /// # Arguments
+    /// 
+    /// * `index` - The index of the vec to insert the item into
+    /// * `value` - The value to insert into the vec
+    ///
+    /// # Reverts
+    /// 
+    /// Reverts if index is larger than length of the vec
+    #[storage(read, write)]
+    pub fn insert(self, index: u64, value: V) {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger than len, there is no space to insert
+        assert(index <= len);
+
+        // for every element in the vec with an index larger than the input index,
+        // move the element up one index.
+        // performed in reverse to prevent data overwriting
+        let mut count = len-1;
```

Comment:
What happens if I call `insert()` on an empty vector (i.e. `len = 0`) and set `index = 0`? Wouldn't `len - 1` cause a problem?

Reply:
then it would underflow and panic right?

Reply:
But this should work, no? Shouldn't Inserting at index `len` (even if `len` is zero) exhibit the same behavior as `push`?

Reply:
Nope, even rust panics here

Reply:
Hmm this is not panicking for me: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c05a852a5b60fb1f13fa37bfd783b1f3

Even then, I think relying on an underflow to get a panic is an antipattern. The better practice is to do all your checking in separate code for safety and readability. The optimizer can then do its magic!

Reply:
Ah sorry my bad I looked at the wrong thing, yeah i guess this works in rust.

After sway starts allowing methods to call other methods ill just add a if len == 0 before the while loop there to early return with a push